### DISCUSSION TITLE
Recommended "Esc" alternative?

### DISCUSSION THREAD

**Discussion by superlou** [helix-editor/helix#661] [Q&A]
On many modern keyboards, the escape key is an awkward reach, and mine definitely pulls me off the home row. In Vim, I typically would use Ctrl-c to return to normal mode (even though there are some functionality differences). In the [remapping docs](https://docs.helix-editor.com/remapping.html), there is an example of mapping Alt-x to move from insert to normal mode. Is that the recommended approach, i.e. does the long-term key strategy leave Alt-x clear, or is it likely to be consumed by some other function?

I didn't see any references to it in the [Keymap Brainstorm](https://github.com/helix-editor/helix/wiki/Keymap-brainstorm) or in any previously closed issues. I apologize if this has been previously litigated.

**Comment by archseer** [comment]
I personally hardware map caps lock to act as ESC on tap and CTRL on hold/in combination with other keys on my keyboards. It turns one of the most useless keys into a really convenient key that reduces strain.

(It also sits right where esc/ctrl historically sat, which explains both why esc and hjkl were used in vi)
![image](https://user-images.githubusercontent.com/1372918/130795671-8513807c-4369-49bf-bd45-501137d67aef.jpeg)

For your custom map, I think alt-x should be fine. You could also consider using right alt?

**Reply by superlou** [reply]
After trying alt-x for a bit. it's not my preference, but I'll keep exploring.

**Reply by pickfire** [reply]
Vim uses ctrl-[

**Reply by archseer** [reply]
Ctrl-[ and ESC are the same terminal code (similar to tab and ctrl-i/o?)

**Reply by superlou** [reply]
Matching VIM seems like a good idea, but the following mapping doesn't seem to work on Windows:

```
[keys.insert]
"C-[" = "normal_mode"
```

As far as being the same terminal code, I definitely don't get the same behavior on the Windows CMD prompt for ctrl-[ as I do for escape.

**Reply by archseer** [reply]
It should just generate an esc keycode without you needing to map anything:

> ESC [..] which can be represented as ASCII code 27 in decimal, Unicode U+001B, or Ctrl+[

At least on Linux/macOS the terminal cannot tell esc and ctrl-[ apart without something like [CSIu](http://www.leonerd.org.uk/hacks/fixterms/) so both keys emit an ESC. No idea about windows.

**Reply by superlou** [reply]
Interesting. I do most of my personal development on Linux at home (where I've been using Vim and ctrl-c), but have been experimenting with Helix mostly in the office on Windows. I never realized there would be different behavior.

Edit: Just confirmed it to myself, that, yep, ctrl-[ just works on Linux.

**Comment by superlou** [comment]
I've heard good things about caps lock, since it's in an ergonomic place on a lot of keyboards. By hardware map, do you mean that you do the mapping outside of applications? Can Helix even see a caps-lock press as an event? If this is more of a discussion, maybe I should move this out of the issue tracker, but I've been trying to bring coworkers into the world of modal-editing, and an app having ergonomic defaults for common keyboards would reduce a friction I've had.

**Reply by eugenesvk** [reply]
> since it's in an ergonomic place on a lot of keyboards

except it's not in an ergonomic place since it requires a lateral pinky move, so the 2nd worst type of finger movement (diagonal is the worst) with the weakest finger on the (generally) weakest hand

**Reply by FedericoCeratto** [reply]
That depends on the keyboard physical layout.

**Comment by archseer** [comment]
I map it either on the OS level, or on my programmable keyboards I change that in the firmware. That way it works in all applications.

https://web.archive.org/web/20141015080124/http://www.economyofeffort.com/2014/08/11/beyond-ctrl-remap-make-that-caps-lock-key-useful

**Comment by superlou** [comment]
That makes a lot of sense for consistency across applications. Unfortunately, in an office where developers don't have admin privileges, it is going to cause some adoption difficulty. I've been trying alt+x for a day or so and feels like a finger twister. I'll play around and see what I can find. Does anyone have favorites?

**Comment by pickfire** [comment]
> I personally hardware map caps lock to act as ESC on tap and CTRL on hold/in combination with other keys on my keyboards. It turns one of the most useless keys into a really convenient key that reduces strain.

Same, I change it system wide. I swap escape and caps (since caps is almost never pressed). As well as left alt and left control (so left thumb for ctrl and right thumb for alt).

> That makes a lot of sense for consistency across applications. Unfortunately, in an office where developers don't have admin privileges, it is going to cause some adoption difficulty. I've been trying alt+x for a day or so and feels like a finger twister. I'll play around and see what I can find. Does anyone have favorites?

Try to get permission is the best choice I think. The admin shouldn't get in your way and let you do stuff in ways that is inconvenient for you.

@archseer Should we have a tips section like this in the wiki? Seemed like quite a few of us have a similar setup (keyboard, japanese, different layout) but people new to helix might not.

Or maybe I can just convert this to discussion?

**Comment by superlou** [comment]
> Or maybe I can just convert this to discussion?

I agree this is definitely more of a discussion (and I missed that this project was using the Discussions feature). I guess the only thing that would elevate it to "issue" is if there is concern that the out-of-the-box experience is hindered by requiring the installation of other tools for ergonomics. Having a compromise back-up might let someone explore what the editor brings to the table and if they want to invest more effort. I think the friendlier selection-verb approach of Helix/Kakoune is a way to introduce more GUI centric users to modal editing, and and an un-ergonomic return-to-normal keycode might make them miss the point.

> The admin shouldn't get in your way and let you do stuff in ways that is inconvenient for you.

I 100% agree with you, but I've definitely been in office environments where the idea of installing an executable that's mission is to read keypresses raises eyebrows.

**Reply by txtyash** [reply]
Wait until you hear about [kmonad](https://github.com/kmonad/kmonad)

**Reply by superlou** [reply]
Agreed that kmonad is really slick, I just can't install it. @WindSoilder has tracked down the reason ctrl-[ doesn't work on Windows to a crossterm bug, so hopefully that can become an expected default.

**Reply by cassepipe** [reply]
You nailed it your first paragraph. I am definitely concerned.

**Comment by zhaojiangbin** [comment]
New to helix here. To avoid reaching for the `ESC` key, I map `A-space` to "normal-mode" with the following in config.toml. However it is only allowed in `[keys.insert]` but not in `[keys.select]`.

> [keys.insert]
> "A-space" = "normal_mode"
>
> [keys.select]
> "A-space" = "normal-mode"

I am getting the following error when hx starts. Does anyone know why?

> Bad config: data did not match any variant of untagged enum KeyTrie for key `keys.select` at line 14 column 1
> Press `<ENTER>` to continue with default config

**Reply by archseer** [reply]
`normal_mode`, not `normal-mode`}

**Reply by zhaojiangbin** [reply]
Oh, what a silly typo I had! I thought the error message was for `keys.select`. 

Thanks.

**Comment by eugenesvk** [comment]
1. For non-admin workflows you can use key chords to exit to normal mode in insert mode like <kbd>i</kbd><kbd>i</kbd> (since I use <kbd>i</kbd> to insert and use the same key to enter/exit other modes, it's nice to be consistent) or <kbd>j</kbd><kbd>j</kbd> (or <kbd>j</kbd><kbd>f</kbd>) to stay on the home row

(use this trick described in this issue https://github.com/helix-editor/helix/issues/2612 to fix one issue with such keybinds)
```toml
[keys.insert] #ⓘ
  'j'	= {f='normal_mode','j'={'f'=[':insert-output /bin/echo -n j','move_char_right','normal_mode']}}
  # 'hij' + 'jf' cleanly exits after typing 'hij' instead of typing an extra jf
  # the char movement was needed because otherwise you end up 1 char left with insert and with append-output at the end of the line you get an insert at the wrong line
```

2. Though an even better alternative (but might require admin to to OS-level rebinds depending on OS) is to use on-hold functionality, e.g. hold🠿<kbd>i</kbd> to exit

Below is a copy&paste of my description of this method from https://github.com/helix-editor/helix/issues/2612#issuecomment-1546955410


Another workaround is to use a good keyboard remapping tool (like Karabiner on a Mac) to implement the workaround I use for the lack of on-hold functionality: basically, make some key like <kbd>i</kbd>
  - insert `i` on __release__ instead of on press
  - simulate some other shortcut on __hold__
  - map ↑ shortcut to enter normal mode in Helix

However, this is limited by:
 - the lack of Helix GUI (so you'd need to use it for the whole terminal as an app filter for remapping)
 - inability of Helix to communicate its mode switches to other apps (so you'd need to have a mechanism that doesn't negatively affect non-insert modes) https://github.com/helix-editor/helix/discussions/6081

(you can also do it for a key chord like <kbd>j</kbd><kbd>f</kbd>, but I think it's more bug-prone due to the aforementioned limitations)

Mac example:
Helix config
```toml
[keys.insert] #ⓘ
  'A-}'	= 'normal_mode' # some unused inconvenient shortcut, ideally free in any mode
```
Goku Karabiner config snippet
```edn
:applications {:wezterm	["^com\\.github\\.wez\\.wezterm$"]}
:main [
 {:des "  🠿i⃣  ⟶ Normal Mode ⇧⎇] (@WezTerm instead of @Helix)" :rules [:wezterm
  [:i         nil nil
   {:alone {:key :i :halt true} :delayed{:canceled[:i]} :held{:key :⇧⎇】 :repeat false}
    :params{:alone 254 :held 255 :delay 255} }]
  ]}
]
```

**Comment by sdoerner** [comment]
Relevant FAQ entry to use "jj" or "jk":
https://github.com/helix-editor/helix/wiki/FAQ#map-jk-or-jj-to-exit-insert-mode

Not sure if you'd ever want to insert "jj" or "jk", but if so, you can still do it with "j<Esc>j" or "j<Esc>k".

**Reply by cassepipe** [reply]
System wide CapsLock/Escape swap (or equivalent) is nice because it unlocks the power of modal editing in cli tools line editors. I doubt any of them would easily support something like `jj` or `jk` even though I could be wrong and even it would suppose choosing a timeout... 
So if you ever want a hx-mode for bash/fish/zsh/gdb and other repls, swapping is much lighter in term of configuration. In Linux  and MacOs it's just a GUI setting or alternatively a shell command and in Windows you just have to modify a register key

**Comment by oxcrow** [comment]
I use `fj` as in `f = {j = ["normal_mode", ":write"]}`

It is *extremely* comfortable to use.
